edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Channels;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using uPLibrary.Networking.M2Mqtt;
+    using uPLibrary.Networking.M2Mqtt.Messages;
+
+    public class MqttBrokerConnector : IMqttBrokerConnector
+    {
+        const int ReconnectDelayMs = 2000;
+
+        readonly IComponentDiscovery components;
+        readonly ISystemComponentIdProvider systemComponentIdProvider;
+        readonly Dictionary<ushort, TaskCompletionSource<bool>> pendingAcks = new Dictionary<ushort, TaskCompletionSource<bool>>();
+
+        readonly object guard = new object();
+
+        Option<Channel<MqttPublishInfo>> publications;
```

Comment:
Why are publications and forwardingLoop options?

Reply:
because when I implemented the auth-endpoint with a similar pattern but using "NULL" instead of none until it is not started, I was told to use Option :)

Reply:
Yes, no nulls pls, but here I think we can init these 2 in the ctor

Reply:
That can be done (by turning the class "not-restartable") - but for me it seems strange for a class that has a method to kick-in its function (the ConnectAsync()), especially the forwarding loop, which is sort of heavy for a constructor.